### ISSUE TITLE
Gradient in keras model with respect to layer outputs is `None` when layer is a Model

### ISSUE THREAD

**Issue by ksabr** [tensorflow/tensorflow#36484]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): centos-release-7-6.1810.2.el7.centos.x86_64

- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:

- TensorFlow installed from (source or binary): Binary

- TensorFlow version (use command below): 2.1.0
- Python version: 3.7.4
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

**Describe the current behavior**
I want to obtain the gradients of a network's output with respect to an intermediate layer which is itself a keras `Model`. This returns `None`. 

**Describe the expected behavior**
It should return something other than `None`.

**Code to reproduce the issue**
```
import tensorflow as tf

input = tf.keras.layers.Input(shape=(28,28))

x = tf.keras.layers.Flatten()(input)
x = tf.keras.layers.Dense(128)(x)
x = tf.keras.layers.ReLU()(x)
x = tf.keras.layers.Dense(10)(x)

small_model = tf.keras.models.Model(inputs=[input], outputs=[x])

y = small_model(input)

y = tf.keras.layers.Dense(20)(y)
y = tf.keras.layers.Dense(30)(y)
y = tf.keras.layers.Dense(10)(y)

big_model = tf.keras.models.Model(inputs=[small_model.input], outputs=[y])

ext_model = tf.keras.models.Model(inputs=[small_model.input], outputs=[big_model.output,  big_model.layers[1].output])

images = tf.random.uniform(shape=(32,28,28))
labels = tf.zeros(shape=(32,))

with tf.GradientTape() as tape:    
    predictions, layer = ext_model(images)

print("grad = ", tape.gradient(predictions, layer))
```

**Other info / logs**
Changing to ```ext_model = tf.keras.models.Model(inputs=[small_model.input], outputs=[big_model.output,  big_model.layers[2].output])``` outputs a gradient different from `None`, which indicates that the problem occurs only when the layer in question is a `Model` itself. But this is precisely the typical use case scenario, where one needs gradients with respect to the outputs of an intermediate layer of a pre-built `tf.keras.applications` model. 

This issue is related (but not identical) to https://github.com/tensorflow/tensorflow/issues/33478 which doesn't seem to have been adequately addressed.

**Comment by amahendrakar** [comment]
Was able to reproduce the issue. Please find the Gist [here](https://colab.sandbox.google.com/gist/amahendrakar/a5327043e792ac7a79e5b78228c68977/36484.ipynb). Thanks!

**Comment by n2cholas** [comment]
You can see the problem more clearly when you visualize the graph on TensorBoard, here's a [gist](https://colab.research.google.com/gist/n2cholas/5dbeaafc9c4d2bbce4e4185bccfe70f4/36484.ipynb). Specifically expand the `ext_model` node. In the `big_model.layers[1].output` case, the `predictions` and `layer` are in two different branches of the graph, so there is no gradient of `prediction` with respect to `layer`. In the `big_model.layers[2].output` case, there is a path between the two nodes, so there is a gradient. This has to do with how `tf.keras` copies the sub-model's ops when you create another model with it.

I'm not sure how to resolve this issue but hopefully this sheds some light on it.

**Comment by jvishnuvardhan** [comment]
@ksabr Is this still an issue? Please close the issue If this was already resolved. Thanks!
